homeassistant/components/seventeentrack/__init__.py

```diff
@@ -1 +1,194 @@
+
+    if unload_ok := await hass.config_entries.async_unload_platforms(
+        config_entry, PLATFORMS
+    ):
+        hass.data[DOMAIN].pop(config_entry.entry_id)
+        if not hass.data[DOMAIN]:
+            hass.services.async_remove(DOMAIN, SERVICE_ADD_PACKAGE)
+            del hass.data[DOMAIN]
+
+    return unload_ok
+
+
+class SeventeenTrackDataCoordinator(DataUpdateCoordinator):
+    """Get the latest data from SeventeenTrack."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        config_entry: ConfigEntry,
+        client: SeventeenTrackClient,
+    ) -> None:
+        """Initialize the data object."""
+        self.hass = hass
+        self.config_entry: ConfigEntry = config_entry
+        self.client = client
+        self.account_id: str = client.profile.account_id
+        super().__init__(
+            self.hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_interval=timedelta(
+                minutes=self.config_entry.options.get(
+                    CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL
+                )
+            ),
+        )
+
+    @property
+    def show_archived(self) -> bool:
+        """Include archived packages when fetching data."""
+        return self.config_entry.options.get(CONF_SHOW_ARCHIVED, DEFAULT_SHOW_ARCHIVED)
+
+    async def _async_update_data(self) -> dict[str, dict]:
+        """Update SeventeenTrack data."""
+        try:
+            packages = await self.client.profile.packages(
+                show_archived=False, tz=str(self.hass.config.time_zone)
+            )
+            summary = await self.client.profile.summary(show_archived=False)
+
+            if self.show_archived:
+                archived_packages = await self.client.profile.packages(
+                    show_archived=True, tz=str(self.hass.config.time_zone)
+                )
+                packages += archived_packages
+
+                archived_summary = await self.client.profile.summary(show_archived=True)
+                for status, qty in archived_summary.items():
+                    summary[status] += qty
+
+        except SeventeenTrackError as err:
+            raise UpdateFailed(f"Error communicating with API: {err}") from err
+
+        _LOGGER.debug("New package data received: %s", packages)
+        _LOGGER.debug("New summary data received: %s", summary)
+
+        return {"packages": packages, "summary": summary}
+
+    async def async_register_service(self) -> bool:
+        """Set up SeventeenTrack."""
+
+        async def async_add_package(service: ServiceCall) -> None:
+            """Add new package."""
+            device_registry: DeviceRegistry = (
+                self.hass.helpers.device_registry.async_get(self.hass)
+            )
+            device = device_registry.async_get(service.data[CONF_DEVICE_ID])
+            if device is None:
+                _LOGGER.error("17Track device not found")
```

Comment:
Nowadays we raise `ValueError` for bad user input and `HomeAssistantError` for api call issues in service handlers. Pass a message as exception argument and it will show in the frontend.